tests/helpers/convert_ipynb_to_script.py

```diff
@@ -0,0 +1,346 @@
+
+        # find the last import statement involving matplotlib, and the first line that uses plt
+        mpl_last_import_index: int = -1
+        mpl_first_usage_index: int = -1
+        for i, line in enumerate(script_lines_new):
+            if "matplotlib" in line and (("import" in line) or ("from" in line)):
+                mpl_last_import_index = i
+
+            if "configure_notebook" in line:
+                mpl_last_import_index = i
+
+            if "plt." in line:
+                mpl_first_usage_index = i
+
+        assert (
+            mpl_last_import_index != -1
+        ), f"matplotlib imports not found! see line {mpl_last_import_index}"
+        if mpl_first_usage_index != -1:
+            assert (
+                mpl_first_usage_index > mpl_last_import_index
+            ), f"matplotlib plots created before import! see lines {mpl_first_usage_index}, {mpl_last_import_index}"
+        else:
+            warnings.warn(
+                "could not find where matplotlib is used, plot disabling might not work!"
+            )
+
+        # insert the cursed things
+        script_lines_new = (
+            script_lines_new[: mpl_last_import_index + 1]
+            + DISABLE_PLOTS["matplotlib"]
+            + script_lines_new[mpl_last_import_index + 1 :]
+        )
+        result_str_TEMP = "\n\n".join(script_lines_new)
+
+    if "circuitsvis" in result_str_TEMP:
+        # find the last import statement involving circuitsvis, and the first line that uses it
+        cirv_last_import_index: int = -1
+        cirv_first_usage_index: int = -1
+
+        for i, line in enumerate(script_lines_new):
+            if "circuitsvis" in line:
+                if (("import" in line) or ("from" in line)) and "circuitsvis" in line:
+                    cirv_last_import_index = i
+                else:
+                    cirv_first_usage_index = i
+
+                if "configure_notebook" in line:
+                    mpl_last_import_index = i
+
+                if "render" in line:
+                    cirv_first_usage_index = i
+
+        assert (
+            cirv_last_import_index != -1
+        ), f"circuitsvis imports not found! see line {cirv_last_import_index}"
+        if cirv_first_usage_index != -1:
+            assert (
+                cirv_first_usage_index > cirv_last_import_index
+            ), f"circuitsvis plots created before import! see lines {cirv_first_usage_index}, {cirv_last_import_index}"
+        else:
+            warnings.warn(
+                "could not find where circuitsvis is used, plot disabling might not work!"
+            )
+
+        # insert the cursed things
+        script_lines_new = (
+            script_lines_new[: cirv_last_import_index + 1]
+            + DISABLE_PLOTS["circuitsvis"]
+            + script_lines_new[cirv_last_import_index + 1 :]
+        )
+        result_str_TEMP = "\n\n".join(script_lines_new)
+
+    return script_lines_new
+
+
+def convert_ipynb(
```

Comment:
I haven't looked at this in detail, but I'm wondering if there's any opportunity to use `nbconvert` to do some of the heavy lifting here? https://nbconvert.readthedocs.io/en/latest/ Always better if we can accomplish the same thing with less new code!

Usage is `jupyter nbconvert --to python notebooks/eval_model.ipynb`

Reply:
Ah I just saw this comment discussing the above https://searchingforsearch.slack.com/archives/C04S0FXBF4P/p1680223916440399?thread_ts=1680221802.003669&cid=C04S0FXBF4P

What was the reason to avoid `nbconvert` here? To avoid installing jupyter in CI?

Reply:
good question! My reasons are as follows: (also adding to #150 for future reference)

- installing jupyter in CI will add a lot of time -- jupyter does many many things, and even `nbconvert` also includes utilities which we dont need (like converting to PDF)
- actually running `nbconvert` is quite slow! on my machine it was ~2sec per notebook. With just 3 notebooks it might not be a huge deal, but hopefully many notebooks from the experiments repo will eventually migrate here
- the most important reason: even if we use `nbconvert`, we would still have to go into the generated file and modify things to prevent plots from being displayed in CI. The file would be half as long if it only included processing notebooks, and shorter still if it didn't have code to process all files in a directory (which would just mean it lives in a different file)